langserver/references.go

```diff
@@ -240,21 +241,54 @@ func sameObj(x, y types.Object) bool {
 	return false
 }
 
-type locationList []lsp.Location
+func sortBySharedDirWithURI(uri string, locs []lsp.Location) {
+	l := locationList{
+		L: locs,
+		D: make([]int, len(locs)),
+	}
+	// l.D[i] = number of shared directories between uri and l.L[i].URI
+	for i := range l.L {
+		u := l.L[i].URI
+		var d int
+		for i := 0; i < len(uri) && i < len(u) && uri[i] == u[i]; i++ {
+			if u[i] == '/' {
+				d++
+			}
+		}
+		if u == uri {
+			// Boost matches in the same uri
+			d++
+		}
+		l.D[i] = d
+	}
+	sort.Sort(l)
+}
+
+type locationList struct {
+	L []lsp.Location
+	D []int
+}
 
 func (l locationList) Less(a, b int) bool {
-	if l[a].URI != l[b].URI {
-		return l[a].URI < l[b].URI
+	if l.D[a] != l.D[b] {
+		return l.D[a] > l.D[b]
+	}
+	if x, y := path.Dir(l.L[a].URI), path.Dir(l.L[b].URI); x != y {
```

Comment:
we could precompute this to make this faster, but the simplicity of this approach makes it more readable. I may in fact remove the precomputation of `l.D` to simplify that code as well if someone things I should.